beacon-chain/cache/committee.go

```diff
@@ -166,3 +201,28 @@ func startEndIndices(c *Committees, index uint64) (uint64, uint64) {
 func key(seed [32]byte) string {
 	return string(seed[:])
 }
+
+func (c *CommitteeCache) checkInProgress(ctx context.Context, seed [32]byte) error {
+	delay := minDelay
+	// Another identical request may be in progress already. Let's wait until
+	// any in progress request resolves or our timeout is exceeded.
+	for {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		c.lock.RLock()
+		if !c.inProgress[key(seed)] {
+			c.lock.RUnlock()
+			break
+		}
+		c.lock.RUnlock()
+
+		// This increasing backoff is to decrease the CPU cycles while waiting
+		// for the in progress boolean to flip to false.
+		time.Sleep(time.Duration(delay) * time.Nanosecond)
+		delay *= delayFactor
+		delay = math.Min(delay, maxDelay)
```

Comment:
We have a few of these identical lines of code across the codebase save for the key type. Instead, we can probably refactor into the top-level cache/ package in another PR and have a keyFn to make this semi-"generic"